### ISSUE TITLE
Literals Union Shortcut Syntax Incompatible with Explicit Unions

### ISSUE THREAD

**Issue by olirice** [python/mypy#8613]
### Summary

A Union containing a Union of Literals (using abbreviated or explicit syntax i.e. Union[Literal["a"], Literal["b"]]` and `Literal["a", "b"]) incorrectly raises an error. See `Pattern3` in the code snippet below for an example.

- Reporting a Bug

```python3
from typing import Dict, Union
from typing_extensions import Literal, Final


###################################
# Test Case: # Union Literal Keys #
###################################
Pattern = Union[
    Dict[Literal["foo"], int],
    Dict[Literal["bar"], int],
    Dict[Literal["bar"], bool],
]

Foo: Literal["foo"] = "foo"
case: Pattern = {Foo: 1}  # Success: no issues found in 1 source file


#####################################
# Test Case: # Combine Literal Keys #
#####################################
Pattern2 = Dict[Literal["foo", "bar"], int]
case2: Pattern2 = {Foo: 1}  # Success: no issues found in 1 source file


####################################################
# Test Case: # Mixed Union & Combined Literal Keys #
####################################################
Pattern3 = Union[
    Dict[Literal["foo", "bar"], int],
    Dict[Literal["baz"], bool],
]
case3: Pattern3 = {
    Foo: 1
}  # Incompatible types in assignment (expression has type "Dict[Literal['foo'], int]", variable has type "Union[Dict[Union[Literal['foo'], Literal['bar']], int], Dict[Literal['baz'], bool]]")
```
* What is the actual behavior/output?
mypy raises error on line 32 (case3) with message 
```
# Incompatible types in assignment (expression has type "Dict[Literal['foo'], int]", variable has type "Union[Dict[Union[Literal['foo'], Literal['bar']], int], Dict[Literal['baz'], bool]]")
```

* What is the behavior/output you expect?
No issues found

* What are the versions of mypy and Python you are using?
python 3.7.2
mypy: 0.770

* Do you see the same issue after installing mypy from Git master?
Yes

* What are the mypy flags you are using? (For example --strict-optional)
None/default

**Comment by olirice** [comment]
Thought of another test case:

```python
Foo: Literal["foo"] = "foo"

Pattern5 = Union[
    Literal["foo", "bar"],
    Literal["baz"]
]

case5: Pattern5 = Foo # Success: no issues found in 1 source file
```
so there must be some interaction with the Dict generic?

**Comment by JelleZijlstra** [comment]
It's because the dict key type is invariant.

**Comment by olirice** [comment]
@JelleZijlstra could you please expand a little on how `Pattern2`  of `Dict[Literal["foo", "bar"], int]` is able to pass if the issue is related to `Dict`'s key type being invariant?

I reread the [documentation on variance of generic types](https://mypy.readthedocs.io/en/stable/generics.html#variance-of-generic-types) and don't follow how it would apply to unions of literals.

I'd really appreciate a nudge in the right direction, reference, or example of how to correctly type something like `Pattern3`

Thanks!

**Comment by JelleZijlstra** [comment]
Because the key type (the first type argument) is invariant, a value of type `Dict[B, Any]` cannot be assigned to a variable of type `Dict[A, Any]` where `B` is a subtype of `A`. In this case, `B` is `Literal["foo", "bar"]` and `A` is `Literal["foo"]`.

As for how to type your example correctly, I don't really understand the use case, but perhaps `TypedDict` would be a better fit for what you're trying to do.
